### DISCUSSION TITLE
Runtime error at ipopt example problem

### DISCUSSION THREAD

**Discussion by omersahintas** [coin-or/CppAD#103] [General]
Dear Bradley,

The execution of [`ipopt_solve_get_started.cpp`](https://github.com/coin-or/CppAD/blob/master/cppad_ipopt/example/get_started.cpp) during `make check` fails with the following error.
```
/usr/local/include/cppad/utility/nan.hpp:188:17: runtime error: division by zero
cppad-20210417 error from a known source:
vector: index greater than or equal vector size
Error detected by false result for
    i < length_
at line 360 in the file 
    /usr/local/include/cppad/utility/vector.hpp
```
 Can you reproduce this error?

**Comment by bradbell** [comment]
What system are you using ?

I am using a Fedora 33 system with the master version
559c841605196d495c3fe1273f44458f24308d27

Using the following comamnds:
```
bin/get_optional.sh
bin/run_cmake.sh --debug_all
cd build
make check_example_ipopt_solve
```
The last command results in the following output:
```
... snip ...
[100%] Linking CXX executable example_ipopt_solve
[100%] Built target example_ipopt_solve
Begin test group example/ipopt_solve
get_started         OK
ode_inverse         OK
retape              OK
memory_leak         OK
All 4 tests passed.
End test group example/ipopt_solve
[100%] Built target check_example_ipopt_solve
```

**Comment by omersahintas** [comment]
I'm using Ubuntu 18.04.4 LTS. I have tried to compile both with gcc/g++ 8.4.0 and gcc/g++ 7.5.0, but got the same result. 

I've further tried linking with both Ipopt 3.13.5 which is the current master head and latest stable release 3.13.3, but that didn't change anything.

**Comment by bradbell** [comment]
I fixed a problem when ipopt was the only optional package installed; see the heading 04-26 on
https://coin-or.github.io/CppAD/doc/whats_new_21.htm#04-26

No the following bash runs the tests in a docker image using ubuntu 18.04:
```
#! /bin/bash -e
cat << EOF > Dockerfile
From dokken/ubuntu-18.04
RUN apt-get update
RUN  DEBIAN_FRONTEND=noninteractive apt-get install -y  \
    cmake \
    git \
    gfortran \
    libblas-dev \
    liblapack-dev \
    pkg-config \
    vim \
    wget
RUN git clone https://github.com/coin-or/cppad.git cppad.git
RUN cd cppad.git && \
    bin/get_ipopt.sh
ENV PKG_CONFIG_PATH='/cppad.git/build/prefix/lib/pkgconfig'
RUN cd cppad.git/build && \
    cmake -D CMAKE_VERBOSE_MAKEFILE=1 -D include_ipopt=true .. && \
    make check_example_ipopt_solve
EOF
docker build --tag ubuntu_cppad .
echo 'test_ipopt.sh: OK'
exit 0
```
I get the following output when I run this script:
```
... snip ...
cd /cppad.git/build/example/ipopt_solve && ./example_ipopt_solve
Begin test group example/ipopt_solve
get_started         OK
ode_inverse         OK
retape              OK
memory_leak         OK
All 4 tests passed.
End test group example/ipopt_solve
... snip ...
```
Please try the corresponding commands on your system; e.g., use `bin/get_ipopt.sh` to install ipopt etc.

**Comment by omersahintas** [comment]
Thanks for your detailed instructions. It's my bad to install ipopt separately and not with the shell script you provided . By using `get_ipopt.sh` I get the correct results.

**Comment by bradbell** [comment]
You can change the install settings in 
https://coin-or.github.io/CppAD/doc/get_ipopt.sh.htm
So that it  installs ipopt with the prefix and version that you want.

You also should be able to use any valid ipopt options using your own script with `bin/get_ipopt.sh` as a staring point.

**Comment by omersahintas** [comment]
Thanks for the hints!

By the way, I see another runtime error when I run the example ipopt problem with gtest. I get 
```
/usr/local/include/cppad/utility/nan.hpp:188:17: runtime error: division by zero

```
But, the optimization works fine and I get the correct results. I didn't encounter this error while executing `make check`.

To inspect this error more in detail, I've set `print_level 10` and have run the test case with gdb, here is the output:
```
Running main() from gtest_main.cc
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from CppAD
[ RUN      ] CppAD.IpoptExample
/usr/local/include/cppad/utility/nan.hpp:188:17: runtime error: division by zero

List of options:

                                    Name   Value                # times used
                         derivative_test = second-order              1
                                max_iter = 10                        1
               point_perturbation_radius = 0                         1
                             print_level = 10                        2
                                      sb = yes                       1
                                     tol = 1e-06                     2
This is Ipopt version 3.13.2, running with linear solver mumps.
NOTE: Other linear solvers might be more efficient (see Ipopt documentation).

Starting derivative checker for first derivatives.

Starting derivative checker for second derivatives.


No errors detected by derivative checker.
(...)
```
So it happens even before the derivative checker. Do you have an idea why this error occurs? I don't think this is something serious.

**Comment by bradbell** [comment]
The file 
    https://github.com/coin-or/CppAD/blob/master/include/cppad/utility/nan.hpp
uses zero / zero to generate a nan for any data type. Using this function was deprecated one 2015-10-04; see
    https://coin-or.github.io/CppAD/doc/nan.htm#nan(zero).Deprecated%202015-10-04
Do you know which test this error is occurring in ?

**Comment by bradbell** [comment]
I am going to change this from a bug report to a discussion.

**Comment by omersahintas** [comment]
To inspect the source of this error, I have tried to catch it with SIGFPE, but that did not work well. I have added an `assert(false)` before [`return zero / zero;`](https://github.com/coin-or/CppAD/blob/403f5737d9718314afd1e10424809a4be83e863d/include/cppad/utility/nan.hpp#L188) and backtraced with gdb:

```
#4  0x0000555555922cf9 in CppAD::nan<double> (zero=@0x7fffffffba60: 0) at /usr/local/include/cppad/utility/nan.hpp:190
#5  0x00005555558b0d3d in CppAD::ipopt::solve_callback<Eigen::Matrix<double, -1, 1, 0, -1, 1>, Eigen::Matrix<CppAD::AD<double>, -1, 1, 0, -1, 1>, (anonymous namespace)::FG_eval>::solve_callback (this=0x61b000000080, nf=1, nx=4, ng=2, xi=..., xl=..., xu=..., gl=..., gu=..., 
    fg_eval=..., retape=false, sparse_forward=false, sparse_reverse=false, solution=...)
    at /usr/local/include/cppad/ipopt/solve_callback.hpp:257
```
I obtain this log when I run [`ipopt_solve_get_started.cpp`](https://github.com/coin-or/CppAD/blob/master/cppad_ipopt/example/get_started.cpp) in gtest and call `CppAD::ipopt::solve<Eigen::VectorXd, OptimizationProblem>(options_, xi_, xl_, xu_, gl_, gu_, optProblem_, solution);`

Maybe UBSan finds an undefined behavior?

**Comment by bradbell** [comment]
I think I have removed the uses of the deprecated version of nan with the following commit:
https://github.com/coin-or/CppAD/commit/fd95f9a13c5f80c5a4769d8e166ef7b361929bc9

**Reply by omersahintas** [reply]
Yes, I can confirm that. That error doesn't occur anymore.

**Comment by omersahintas** [comment]
If the division in [`div.hpp`](https://github.com/coin-or/CppAD/blob/fd95f9a13c5f80c5a4769d8e166ef7b361929bc9/include/cppad/core/div.hpp#L23) returns a nan, how is it treated internally? Are there rules to catch certain types of exceptions?

**Comment by bradbell** [comment]
Perhaps the following examples make this clear ?
https://github.com/coin-or/CppAD/discussions/89#discussioncomment-276440
https://coin-or.github.io/CppAD/doc/cond_exp.cpp.htm
